docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,312 @@
+  dependency system automatically instantiates bears, which might lead to
+  unwanted side-effects).
+
+Official support for virtual files
+----------------------------------
+
+IDEs like IntelliJ use virtual files to represent files in a filesystem (VFS)
+and perform operations on them. Hence NextGen-Core provides official support
+for virtual files. Bears have to point to the right file data objects when
+run, whether they are real files or virtual ones. This makes coala easier to
+integrate with IDEs.
+
+Task Objects
+------------
+
+These are the representation of tasks performed by bears. Structure-wise
+they are a tuple containing the tuple of positional arguments and a dict of
+keyword arguments to the ``analyze`` (even to the ``execute_task``) function.
+The task objects are instrumental to the caching mechanism as their hash values
+are stored in the cache along with the bear results and are looked up during
+each coala run to fetch the results. Lets say we have a FileBear that looks like
+this:
+
+::
+    class SomeFileBear(FileBear):
+
+    def analyze(self, file, filename, filename_prefix: str='',
+                filename_suffix: str=''):
+        yield "Some kind of analysis"
+
+Its corresponding task object would like the one below.
+
+::
+
+    [
+        [(file, filename), {'filename_prefix': "", 'filename_suffix': ""}],
+    ]
+
+Improved dependency system
+--------------------------
+
+The NextGen-Core introduces a better dependency management system than the
+one used by the old core. It features the following improvements:
+
+* A class ``DependencyTracker`` manages this task of dependency management.
+  Dependencies are added and resolved by this class and it checks for
+  circular dependencies.
+* Dependency relations between two objects are tracked using a directed
+  graph. When two objects (or nodes) are connected with a directed edge they
+  form a dependency relation.
+
+Eventhough bears still have to pass ``Result`` instances to communicate with
+coala, it is now possible to pass arbitrary python objects. This has its
+benefits in case of dependency bears since they can pass any arbitrary
+object instead of just ``Result`` objects.
+
+``DependencyBear``
+------------------
+
+Handling of bear dependencies by the old core wasn't effective. The
+old core used a queuing mechanism to communicate between bear runs and passed
+all the dependencies to the ``dependency_results`` argument of the ``run``
+function which led to problems of a ``GlobalBear`` not running if any one of its
+dependencies (e.g. a ``local bear``) terminated, the NextGen-Core improves on
+this.
+
+A new bear known as the ``DependencyBear`` is also provided by the NextGen-Core
+making it more convenient for bear developers to write dependency bears, by
+passing the dependency results using task objects. This technique of handling
+dependencies make it possible for the ``DependencyBear`` to support caching.
```

Comment:
Hmm maybe mentioning tasks here is too much of implementation details. Not sure on that one yet, will reread it again once the other comments are fixed.